### PR TITLE
♻️ Abstract away aspect ratio from grid-layer to be used by page-layer

### LINKED ISSUE #31772: [Story-page-layer] Abstract away funtionality of grid-layer into base-layer
We want to reuse some of the grid-layer functionality into page-layers, so we need to move the functionality into base-layer and rename the CSS classes to a more general name

### PR DESCRIPTION
**Pull request by mszylkowski**
Closes #31772 

Changes the template classes from `grid` to `layer` so they can be reused and controlled by `amp-story-base-layer`.

Moved to `amp-story-base-layer`:
- page size listeners
- aspect-ratio variables update
- template classes processing
- function to check if page should be prerendered (either first page or page in hashString)

All the functionality abstracted away does not add any behaviors to subclasses of base-layer directly, they just provide the methods that can be called if subclasses want them.

Prototype: https://github.com/ampproject/amphtml/pull/31751

### CODE REVIEW

**Comment by gmajoulet** [comment]
Do you have the public API and design documented and explained somewhere, so we can all see where this is going?

**Comment by newmuis** [comment]
Do we actually want to port these over?  I would expect that they're low usage and not particularly helpful to me the current state of the ecosystem.

**Comment by mszylkowski** [comment]
Good point, we can discuss more of the approach offline (don't have the full details on GH yet). I assume the question is about template classes, and I'd say the `fill` one may be useful but the other ones can be disabled. `fill` helps a lot with images and videos to make them take the whole layer that it'd be good to keep (unless we provide an alternative). I agree the other ones (`thirds`, `vertical`, `horizontal`) wouldn't be used much and can be invalid for `amp-story-page-layer` components.

**Comment by mszylkowski** [comment]
Stuck on not passing Travis because of #31756. @Enriqe is OOO this week but he has more context on that

**Comment by mszylkowski** [comment]
@newmuis removed the templates from the base-layer (which made the PR a lot more concise), but since we talked about using `layout="fill"` instead of the templates I had to add the fullscreen ad support by doing that along with `amp-story-grid-layer[template="fill"] > amp-ad`. Haven't seen other places where it's needed to account for that yet (will revise later on!)

**Review by gmajoulet — APPROVED** [review]

File: extensions/amp-story/1.0/amp-story-base-layer.js [hunk 1]
```diff
@@ -38,4 +48,66 @@ export class AmpStoryBaseLayer extends AMP.BaseElement {
...
+      selector += `, amp-story-page#${escapeCssSelectorIdent(hashId)}`;
+    }
+    const selectorNodes = this.win.document.querySelectorAll(selector);
+    this.isPrerenderActivePage_ =
+      selectorNodes[selectorNodes.length - 1] === this.element.parentElement;
+    return this.isPrerenderActivePage_;
+  }
+
+  /** @protected */
+  initializeAspectRatioListeners() {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-base-layer.js:73]: Let's make this method more general about aspect ratio things (presets, aspect-ratio, scaling-factor...).
Each layer (grid and page) should override it and call super(). The page-layer will read different attributes (preset) but still run the same logic, it'd be great to plan for it in this PR.

**Comment by mszylkowski** [comment]
Closing since most of this will be refactored into grid-layer

**Pull request closed without merging**